### ISSUE TITLE
NaN after large number of iterations on 200d GloVe training

### ISSUE THREAD

**Issue by moscow25** [stanfordnlp/GloVe#15]
Thanks for a great codebase! 

When training a word-GloVe on a 100M tweets dataset with 30 minimum word-occurrence and resulting in a 1.2M effective vocabulary size, the GloVe trains great, albeit slowly. The error decreases at each iteration... and then after ~25 iterations, it suddenly goes NaN. Will follow up a log.

This happens for me every time, after slightly different iteration count. I don't think it is a memory problem. Rather, perhaps it's some kind of division by a very very small number, that becomes NaN. I'm thinking tiny, non-zero gradients, or taking the inverse of that gradient, resulting in a sneaky divide-by-zero. 

If my theory had some merit, I would appreciate a code pointer, where I might look for the NaN in the gradient descent?

Again, it runs for hours, on a large 1.2M vocabulary. Great GloVe... then sudden a NaN. If I choose a much higher word occurrence cutoff or a smaller dimension like 50d, I can not reproduce the error. But with a 1.2M vocabulary and 15-word occurrence window... after running long enough, it happens every time.

Thanks! Love the GloVe.

**Comment by moscow25** [comment]
./demo.sh 
mkdir -p build
gcc src/glove.c -o build/glove -lm -pthread -Ofast -march=native -funroll-loops -Wno-unused-result
gcc src/shuffle.c -o build/shuffle -lm -pthread -Ofast -march=native -funroll-loops -Wno-unused-result
gcc src/cooccur.c -o build/cooccur -lm -pthread -Ofast -march=native -funroll-loops -Wno-unused-result
gcc src/vocab_count.c -o build/vocab_count -lm -pthread -Ofast -march=native -funroll-loops -Wno-unused-result
BUILDING VOCABULARY
Processed 1755844961 tokens.
Counted 22289076 unique words.
Truncating vocabulary at min count 25.
Using vocabulary of size 1193516.

COUNTING COOCCURRENCES
window size: 10
context: symmetric
max product: 13752509
overflow length: 38028356
Reading vocab from file "vocab.txt"...loaded 1193516 words.
Building lookup table...table contains 171365725 elements.
Processed 1755844961 tokens.
Writing cooccurrences to disk...........125 files in total.
Merging cooccurrence files: processed 924208891 lines.

SHUFFLING COOCCURRENCES
array size: 255013683
Shuffling by chunks: processed 924208891 lines.
Wrote 4 temporary file(s).
Merging temp files: processed 924208891 lines.

TRAINING MODEL
Read 924208891 lines.
Initializing parameters...done.
vector size: 200
vocab size: 1193516
x_max: 10.000000
alpha: 0.750000
iter: 001, cost: 0.159467
iter: 002, cost: 0.123287
iter: 003, cost: 0.103016
iter: 004, cost: 0.091396
iter: 005, cost: 0.084560
iter: 006, cost: 0.080124
iter: 007, cost: 0.076975
iter: 008, cost: 0.074624
iter: 009, cost: 0.072742
iter: 010, cost: 0.071216
iter: 011, cost: 0.069942
iter: 012, cost: 0.068856
iter: 013, cost: 0.067923
iter: 014, cost: 0.067107
iter: 015, cost: 0.066391
iter: 016, cost: 0.065768
iter: 017, cost: 0.065206
iter: 018, cost: 0.064694
iter: 019, cost: 0.064241
iter: 020, cost: 0.063825
iter: 021, cost: 0.063450
iter: 022, cost: 0.063102
iter: 023, cost: 0.062777
iter: 024, cost: 0.062470
iter: 025, cost: 0.062172
iter: 026, cost: 0.061888
iter: 027, cost: 0.061652
iter: 028, cost: nan
iter: 029, cost: nan
iter: 030, cost: nan
capital-common-countries.txt:
ACCURACY TOP1: 0.00% (0/506)
capital-world.txt:
ACCURACY TOP1: 0.00% (0/3629)
currency.txt:
ACCURACY TOP1: 0.00% (0/752)
city-in-state.txt:
ACCURACY TOP1: 0.00% (0/2467)
family.txt:
ACCURACY TOP1: 0.00% (0/506)
gram1-adjective-to-adverb.txt:
ACCURACY TOP1: 0.00% (0/930)
gram2-opposite.txt:
ACCURACY TOP1: 0.00% (0/702)
gram3-comparative.txt:
ACCURACY TOP1: 0.00% (0/1332)
gram4-superlative.txt:
ACCURACY TOP1: 0.00% (0/1122)
gram5-present-participle.txt:
ACCURACY TOP1: 0.00% (0/1056)
gram6-nationality-adjective.txt:
ACCURACY TOP1: 0.00% (0/1445)
gram7-past-tense.txt:
ACCURACY TOP1: 0.00% (0/1560)
gram8-plural.txt:
ACCURACY TOP1: 0.00% (0/1332)
gram9-plural-verbs.txt:
ACCURACY TOP1: 0.00% (0/870)
Questions seen/total: 93.17% (18209/19544)
Semantic accuracy: 0.00%  (0/7860)
Syntactic accuracy: 0.00%  (0/10349)
Total accuracy: 0.00%  (0/18209)

**Comment by igorbrigadir** [comment]
I can't offer a detailed explanation (I'm not 100% sure why that worked in a similar setting for me) but try setting a different (lower) alpha with the same number of iterations.

**Comment by moscow25** [comment]
Thanks Igor, but this does not address my question. Where in the code could I be getting a NaN? Of course I can avoid it with hyper parameter selection. But I would like to find and fix the problem if possible, so I could train massive GloVes in high dimensions, on sparse-ish data, for a long period of time.

And perhaps so that others could do the same.

Sincerely,
Nikolai

**Comment by igorbrigadir** [comment]
Absolutely! Much better to actually fix it rather than avoiding it with parameters!

**Comment by moscow25** [comment]
I would like to find and fix it. Do you have any pointers in the code, where to look? 

Thanks,
Nikolai

**Comment by Russell91** [comment]
The code throwing the NaNs is certainly in src/glove.c. See the train_glove() function on line 228 in particular. I recommend you reduce the number of threads to 1 before debugging, and perhaps even replace the pthread calls for standard function calls before debugging. It should be possible to include assertion statements to the effect that no NaNs are produced in various places and then debug with gdb. Given the fact that this indeed likely affects many other users, if you (or anyone else with this problem) are able to provide a corpus, I'll take a look at what's going on.

**Comment by moscow25** [comment]
Thanks! I was thinking something similar so good to hear this from you.

Going to take a look tonight and respond in some details. Not sure I can share this corpus... at least not easily. Let me see about that also.

Best,
Nikolai

**Comment by moscow25** [comment]
Hey @Russell91, suppose I had a theory that line 106 could in some case get into a NaN.
https://github.com/stanfordnlp/GloVe/blob/master/src/glove.c#L106

I'm not quite the C-wizard. What's the easiest way to test that this computation does not yield NaN, and if it does, printf() some stuff? Will figure it out if I don't hear back, but thought maybe can save a bunch of time. If I can catch and replicate it, will do my best to make it reproducible for you, as well. Currently, it takes a couple of hours to fail, but pretty consistent at which point it fails when it does. 

My theory is that something falls over and creates a numerical divide by zero, when dividing by a very small, non-zero value. I've see this before in a similar adaptive gradient descent problem. 

Good point about running in one thread to test. Thinking about how to change parameters to try to get a fail to happen sooner. 

Best,
N

**Comment by moscow25** [comment]
Pretty sure I solved it. Running the 7-hour job on my dataset now to be sure. 

Will send a review tonight if it's confirmed. Basically just need to check for isnan() and isinf() before applying the gradient update. If error, just skip the update and keep going.

For a smaller set, I even introduced random NaNs into the gradient, and it trained to the same number.

**Comment by Russell91** [comment]
Wow, sounds very useful if it indeed works.

**Comment by moscow25** [comment]
So it does appear to work. Adding the checks, I'm able to train for more iterations, and to a lower loss in the co-occurence metric. I also tested it by faking bad gradients, and the network is able to skip those updates and train to the same number, as well.

Performance does not seem slower, but I did not carefully benchmark it. 

Would you like me to send my change a code review?

**Comment by Russell91** [comment]
Yea, a significant speed decrease would have been unexpected from adding an inline if ... then ... check. Please do send a pull request. If you haven't done that before, just 1) hit the fork button on the glove repo 2) add your own repo locally with `git remote add moscow https://github.com/moscow25/` or such, 3) push your changes locally e.g. `git push moscow master`, and then come back to stanfordnlp/glove and hit the create pull request button.

**Comment by moscow25** [comment]
Excellent. I'll do that tonight. Have sent pull requests before... but it's
been a while :-)

**Comment by moscow25** [comment]
Pull request created. Thanks! Let me know if you want style changes, or modify them yourself as needed. Thanks for a nice repo.

**Comment by Russell91** [comment]
Looks great! The issue should be addressed by https://github.com/stanfordnlp/GloVe/pull/16.
